touchforms/formplayer/static/formplayer/script/fullform-ui.js

```diff
@@ -1,634 +1,459 @@
-        var k = +vs[i];
-        if (k != 0) {
-          meta.as_single.push(k);
+    var meta = {};
+
+    if (type == "date") {
+        meta.mindiff = style.before != null ? +style.before : null;
+        meta.maxdiff = style.after != null ? +style.after : null;
+    } else if (type == "int" || type == "float") {
+        meta.unit = style.unit;
+    } else if (type == 'str') {
+        meta.autocomplete = (style.mode == 'autocomplete');
+        meta.autocomplete_key = style["autocomplete-key"];
+        meta.mask = style.mask;
+        meta.prefix = style.prefix;
+        meta.longtext = (style.raw == 'full');
+    } else if (type == "multiselect") {
+        if (style["as-select1"] != null) {
+            meta.as_single = [];
+            var vs = style["as-select1"].split(',');
+            for (var i = 0; i < vs.length; i++) {
+                var k = +vs[i];
+                if (k != 0) {
+                    meta.as_single.push(k);
+                }
+            }
         }
-      }
     }
-  }
 
-  if (type == "select" || type == "multiselect") {
-    meta.appearance = style.raw;
-  }
+    if (type == "select" || type == "multiselect") {
+        meta.appearance = style.raw;
+    }
 
-  return meta;
+    return meta;
 }
 
-function Form(json, adapter) {
-  this.adapter = adapter;
-  this.children = [];
-  this.template = window.JST[JST_BASE_DIR + 'fullform-ui/form.html'];
-
-  this.init_render = function() {
-    this.$container = $(this.template());
-    this.$title = this.$container.find('#title');
-    this.$children = this.$container.find('#form');
-
-    this.$title.text(json.title);
-    render_elements(this, json.tree);
+/**
+ * Base abstract prototype for Repeat, Group and Form. Adds methods to
+ * objects that contain a children array for rendering nested questions.
+ * @param {Object} json - The JSON returned from touchforms to represent the container
+ */
+function Container(json) {
+    var self = this;
+    self.fromJS(json);
+
+    /**
+     * Used in KO template to determine what template to use for a child
+     * @param {Object} child - The child object to be rendered, either Group, Repeat, or Question
+     */
+    self.childTemplate = function(child) {
+        return ko.utils.unwrapObservable(child.type) + '-fullform-ko-template';
+    }
+}
 
-    var form = this;
-    this.$container.find('#submit').click(function() {
-        var proceed = adapter.presubmitfunc();
-        if (!proceed) {
-          return;
+/**
+ * Reconciles the JSON representation of a Container (Group, Repeat, Form) and renders it into
+ * a knockout representation.
+ * @param {Object} json - The JSON returned from touchforms to represent a Container
+ */
+Container.prototype.fromJS = function(json) {
+    var self = this;
+    var mapping = {
+        caption: {
+            update: function(options) {
+                return options.data ? DOMPurify.sanitize(options.data.replace(/\n/g, '<br/>')) : null;
```

Comment:
this is actually pretty nice. we no longer have to worry about the whole `allow_html` flag. this library purifies the string by only allowing safe html to be in the caption